server/src/services/asset.service.ts

```diff
@@ -464,4 +493,95 @@ export class AssetService extends BaseService {
       await this.jobRepository.queue({ name: JobName.SidecarWrite, data: { id, ...writes } });
     }
   }
+
+  async getAssetEdits(auth: AuthDto, id: string): Promise<AssetEditsDto> {
+    await this.requireAccess({ auth, permission: Permission.AssetRead, ids: [id] });
+    const edits = await this.assetEditRepository.getEditsForAsset(id);
+    return {
+      assetId: id,
+      edits,
+    };
+  }
+
+  async editAsset(auth: AuthDto, id: string, dto: EditActionListDto): Promise<AssetEditsDto> {
+    await this.requireAccess({ auth, permission: Permission.AssetEdit, ids: [id] });
+
+    if (dto.edits.length === 0) {
+      throw new BadRequestException('At least one edit action must be provided');
+    }
+
+    const asset = await this.assetRepository.getById(id, { exifInfo: true });
+    if (!asset) {
+      throw new BadRequestException('Asset not found');
+    }
+
+    if (asset.type !== AssetType.Image) {
+      throw new BadRequestException('Only images can be edited');
+    }
+
+    if (asset.livePhotoVideoId !== null) {
+      throw new BadRequestException('Editing live photos is not supported');
+    }
+
+    if (isPanorama(asset)) {
+      throw new BadRequestException('Editing panorama images is not supported');
+    }
+
+    if (asset.originalPath?.toLowerCase().endsWith('.gif')) {
+      throw new BadRequestException('Editing GIF images is not supported');
+    }
+
+    // verify there are unique actions
+    // mirror can be duplicated but must have different parameters
+    const actionSet = new Set<string>();
+    for (const edit of dto.edits) {
+      const key = edit.action === EditAction.Mirror ? `${edit.action}-${JSON.stringify(edit.parameters)}` : edit.action;
+      if (actionSet.has(key)) {
+        throw new BadRequestException('Duplicate edit actions are not allowed');
+      }
+      actionSet.add(key);
+    }
+
+    // check that crop parameters will not go out of bounds
+    const { width: assetWidth, height: assetHeight } = getDimensions(asset.exifInfo!);
+
+    if (!assetWidth || !assetHeight) {
+      throw new BadRequestException('Asset dimensions are not available for editing');
+    }
+
+    const crop = dto.edits.find((e) => e.action === EditAction.Crop)?.parameters;
+    if (crop) {
+      const { x, y, width, height } = crop;
+      if (x + width > assetWidth || y + height > assetHeight) {
+        throw new BadRequestException('Crop parameters are out of bounds');
+      }
+    }
+
+    await this.assetEditRepository.storeEdits(id, dto.edits);
+    await this.jobRepository.queue({
+      name: JobName.AssetGenerateThumbnails,
+      data: { id, source: 'edit', notify: true },
+    });
+
+    // Return the asset and its applied edits
+    return {
+      assetId: id,
+      edits: dto.edits,
```

Comment:
If you are going to return edits, you should return edits from the database (return them on line 560 and then return them here). We should not just echo back the input, which might have been saved differently in the database, depending on any data transformations that happened, or any generated fields that might have a default value that was originally supplied.